PR: Add file copy-paste feature to project explorer

Khalilsqu opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)
* [x] Included a screenshot (if affecting the UI)

<!--- Explain what you've done and why --->
This PR adds some necessary and highly desirable modern functionality to Spyder-IDE. 
First, it is now possible to simply copy a single or multiple files’ absolute or relative paths to clipboard. This will help in cases where getting files names as inputs parameters for various functions e.g. pandas input csv files etc.
Below is a demonstration how it works for both absolute and relative paths.

![path_1](https://user-images.githubusercontent.com/22060413/50390971-d7ff3a80-0735-11e9-885b-3e3f6956ed86.gif)


Additionally, it is also now possible to copy and paste files in the project explorer or file explorer. This will close the issue #8321. I believe this feature will ease and save developers time while working and managing projects files. 
Below is a demonstration of it works. 

![copy_save_file_2](https://user-images.githubusercontent.com/22060413/50390972-dc2b5800-0735-11e9-8f64-e6fd7d3af4ca.gif)

The files which are saved in the clipboard can be pasted in external file system. Also, external clipboard can be saved in Spyder project explorer tree.

Demo Below.

![external_3](https://user-images.githubusercontent.com/22060413/50390975-e2213900-0735-11e9-9d3d-aae3e539a308.gif)

Not requiring to open QPopUp menus and messages I believe is preferred  for simplicity and faster code editing.

I am submitting this PR for review, I thinks there are some major improvement can be made into these functionalities. For example:
1.	Implementing keyboard shortcuts.
2.	Adding icons.
Thanks again and I am looking forward to your opinions. 
If you find these new features complicating usability and interface of Spyder, you can then close this PR.

Regards


### Issue(s) Resolved 

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #2641 #8321


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @Khalilsqu - Khalil Al Hooti 

<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @Khalilsqu! Thanks for submitting the PR.

 - In the file [`spyder/plugins/explorer/widgets.py`](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py), following are the PEP8 issues :

> [Line 392:17](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L392): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 393:56](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L393): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 394:17](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L394): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 395:57](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L395): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 396:17](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L396): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 398:17](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L398): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 420:50](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L420): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 880:44](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L880): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 896:72](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L896): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 896:76](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L896): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 897:41](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L897): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 901:42](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L901): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 902:1](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L902): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 911:17](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L911): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 922:65](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L922): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 923:37](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L923): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 923:78](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L923): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets
> [Line 924:37](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L924): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 955:69](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L955): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 956:25](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L956): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 956:55](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L956): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets
> [Line 959:65](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L959): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 960:25](https://github.com/spyder-ide/spyder/blob/656babc66e07163cd1bbcfcd60199a2eaff21d71/spyder/plugins/explorer/widgets.py#L960): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

CAM-Gerlach commented:
@Khalilsqu Thanks so much for these changes; they are much appreciated! In addition to closing #8321 , the first part of your changes also is exactly what is requested in #2641 . They are indeed very warmly received.

I'll have a full review shortly, but make sure to add #2641 next to ``Fixes #`` so it closes that issue as well, and please type your name next to the "I affirm this is true and correct: " statement to certify we have permission to use your great work (you still retain full copyright; its not a CLI). Thanks again!

Khalilsqu commented:
Thank you very much @CAM-Gerlach . Your review and feedback indeed is very constructive. Your all point are valid and certainly well help in making this PR successful and more likely to be accepted and merged. 
For the time being I will make slight modification, like pep correction and undoing changes to lines not belonging my commits. 
I will be having limited access to a PC till a week after the new year. However, I will take care of all your suggestion and recommendation after this period or whenever I get change if sooner.  

Thank you very much.

Khalilsqu commented on spyder/plugins/explorer/widgets.py:
```diff
@@ -388,23 +388,38 @@ def create_file_manage_actions(self, fnames):
+        copy_relative_path_action = create_action(self, 
+                _("Copy Relative Path"), triggered=self.copy_relative_path)
+        copy_file_clipboard_action = create_action(self, 
+                _("Copy to Clipboard"), triggered=self.copy_file_clipboard)
```
I made this change, but I actually think it is better to use short names. I will include the shortcuts after the name. similar to vsc. The pop window will get too large. However, if you find it fine, I will go with your suggestions.

example from vscode
![image](https://user-images.githubusercontent.com/22060413/50413987-97252580-080a-11e9-8be3-a60236382359.png)

CAM-Gerlach replied:
Sorry, but I find that even less clear than before. Our priority for UI text should be clarity and explictness above all, and remember the Zen of Python: ``Explicit is better than implicit``. At least, without a well-articulated, evidence-based reason.

Therefore, please revert it to ``Copy File to Clipboard``, ``Copy Absolute Path``, and ``Copy Relative Path``—and where did ``Paste File from Clipboard`` go?

Also, does ``Ctrl-K`` come from another editor/IDE? VSCode? If so, which command? For simplicity/conciseness, unless it is a very common shortcut used in multiple editors/IDEs, we should just keep it at one hotkey by default, and we'll have shortcut profiles for other editors/IDEs in Spyder 4 so users will be able to easily switch to one emulating their favorite one.

Khalilsqu replied:
Sorry for not making it clear. The image above actually is from vscode not my implementation for these new functionalities! I am agreeing to your suggestions. 

Indeed giving the user the ability to change modify keyboard shortcuts to their preference is a great advantage.

Khalilsqu commented:
![spyder4](https://user-images.githubusercontent.com/22060413/50426480-11df5680-0886-11e9-9adb-dd9e5431e8a9.gif)

1. Added keyboard shortcut to each functionality using the suggested keyboard shortcuts.
2. made file naming system easier. It is similar to file copying system available in vscode. New copied file incremented by one.
3. Added some icons to two functionalities (copying and pasting files). I could not find suitable one for copying absolute and relative paths.
4. Added the keyboard shortcuts names into the UI. still not looking good!! I will try to make them tabulated and keyboard shortcuts right justified. 
5. made copied files absolute and relative names follow a similar standard in Windows as is the case for Linus using forslash as a seperated instead of back slash.


Next Step>
1 . I will add some test functionalities as requested.
2 . Correct the appearance keyboard shortcuts in the GUI.

Khalilsqu commented:
> Thanks so much @Khalilsqu ; really looking forward to seeing your further changes and this getting implemented in Spyder. Let us know if you need guidance on any of the larger changes. Best!

Thank you very much for your recommendations and reviews. Highly appreciated.

Could you tell me why I am getting old commits from other people in this PR, and why PEP shows errors from lines of code I have not touched? Sorry if my questions  are stupid, I have been Matlab used and recently switched to Python and open source community.

By the way could you help me writing my PhD!!, Just kidding. I noticed you quite often explain thing in very great detail and with passion even if the question are dump (e.g. some questions in gitter). I think your are very help to new users of Spyder or even new contributes. Cheers!

CAM-Gerlach commented:
Thanks so much for your improvements! I will test the functionality and have another review for you shortly. If you could fix the extra commits being shown by e.g. rebasing on ``master``, that would make it much easier for me to do so, thanks.

> Relative path is defined as the the relative to the current working directory.

Is this different from before? It should be relative to the output of ``get_cwd_or_home()`` which I thought you already had.

> 2 . Correct the appearance of keyboard shortcuts in the GUI.

Yes, see my forthcoming review for how to do this correctly.

> Could you tell me why I am getting old commits from other people in this PR, and why PEP shows errors from lines of code I have not touched? 

I'm not sure without knowing exactly what ``git`` commands you executed but something went wrong there. You should be able to fix it by ``git pull`` ing the latest ``master``, switching to ``add-copy-paste-project-explorer``, and doing ``git rebase master add-copy-paste-project-explorer``, then force pushing with ``git push -f``.

> Sorry if my questions are stupid,

They are not stupid at all! You probably know more than me already, haha!

> I think your are very help to new users of Spyder or even new contributes. Cheers!

Thanks so much; I'm really glad to hear that! <3

Khalilsqu commented:
I closed the PR. I moved away from my University PC to home PC and then forked my remote repo into my home pc. Whenever, I try to push, git asks me to create a new PR. I almost spent whole day trying to push to the already existing PR without success.

Khalilsqu commented:
Sorry about that, I am not sure what was going wrong, I cloned my forked repo from Github and edited the files in my local machine. When I pushed changes to remote repo it always asked me to open a new PR. My deepest apologies. after some tried and error since it was always asking me to open a new repo and since I was already having a forked repo, I decided to deleted it an make a new fork and start again. Currently, I am trying to write to some tests on these new functionalities. 

Thank you very much for your time.

CAM-Gerlach commented:
> I cloned my forked repo from Github and edited the files in my local machine. 

Right, exactly. As long as you cloned your forked repo, checked out your feature branch pushed to the ``add-copy-paste-project-explorer`` branch on your fork, you were doing everything correctly already.

> When I pushed changes to remote repo it always asked me to open a new PR.

This is normal; Github always does this (as of perhaps a month or so ago) when you push changes to a non-default branch (typically ``master``) of any repo. However, I'm not sure if it still does this if you push to a branch that is already connected to a PR; if that's the case (I can't remember) its possible you were pushing your changes to a different branch than the one connected to this PR; the solution in that case would just be to push to the correct branch on ``origin``. If not then there was nothing to worry about so long as the changes showed up on the PR within a minute or two.

> I decided to deleted it an make a new fork and start again.

If you did the three steps above correctly (clone from fork, checkout feature branch, push to feature branch on your fork) then this was your only mistake, which perma-nuked this whole PR. At the very least, if you confirmed commits really weren't showing up on this PR for some reason, you could have closed it and opened a new one with the same or a different branch; then I could investigate the problem (if there was one) and if it was solvable the situation on this PR would at least be recoverable.

> Currently, I am trying to write to some tests on these new functionalities.

Great! Let us know if you need help and guidance.

Khalilsqu commented:
Thanks for your help.

This closed PR is reopened in #8494
